### ISSUE TITLE
Bug in multipart POST

### ISSUE THREAD

**Issue by gekomad** [http4s/http4s-jdk-http-client#576]
In version 0.3.7, multipart POST does not seem to work. Using the standard http4s client everything is fine.
    
```scala
    val multipart = Multipart[IO](Vector(Part.formData("form", """foo""")))
    
    val request: Request[IO] = Request[IO](POST, url).withEntity(multipart).withHeaders(multipart.headers)
```

```scala
    // http4s standard client works fine

    BlazeClientBuilder[IO](global).resource.use { client0 =>
        client0.expect[String](request)
    }
```

```scala
    // http4s-jdk-http-client client 0.3.7  doesen't work

    val baseClient = HttpClient.newBuilder().version(HttpClient.Version.HTTP_2)
    val client0: Client[IO] = JdkHttpClient(baseClient.build())
    client0.expect[String](request)
```

**Comment by amesgen** [comment]
I can't reproduce this against https://httpbin.org/post. Can you reproduce this with a publicly available site? What error do you get? Can you ensure that this also happens when using `java.net.http.HttpClient` directly?

**Comment by gekomad** [comment]
Unfortunately I'm calling a IIS private server and I can't debug it, the error is 400 Bad Request. sttp client works too.
Do you know a public debug service for IIS?

**Comment by amesgen** [comment]
Sorry, I am not aware of such a site. Can you ensure that this also happens when using `java.net.http.HttpClient` directly? http4s-jdk-http-client is a light wrapper around that, so this would be helpful to debug the issue.

**Comment by gekomad** [comment]
Using `java.net.http.HttpClient` works. I used this code `https://urvanov.ru/2020/08/18/java-11-httpclient-multipart-form-data/`

**Comment by gekomad** [comment]
I wrote two clients one entirly in java with `java.net.http.HttpClient` and one in scala using http4s-blaze-client.
I sent a multipart POST with field `foo` to https://httpbin.org/post; in response of java version I see the filed `foo` but in scala version I don't see that field.

```
"form": {
    "foo": "{\"aaaaa\" : 123 }"
  }, 
```

--------------------- CLIENT JAVA ------------------------------------

```java
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublisher;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class HttpClientMultipartExample {

    public static void main(String[] args) throws IOException,
            InterruptedException, URISyntaxException {

        String url = "https://httpbin.org/post";

        String boundary = "-------------oiawn4tp89n4e9p5";
        Map<Object, Object> data = new HashMap<>();

        data.put("foo", "{\"aaaaa\" : 123 }");

        BodyPublisher x = oMultipartData(data, boundary);
        HttpRequest request = HttpRequest.newBuilder().uri(new URI(url))
                .headers("Content-Type", "multipart/form-data; boundary=" + boundary).POST(x).build();

        HttpClient client = HttpClient.newHttpClient();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        System.out.println("response  = " + response);
        System.out.println("response statusCode = " + response.statusCode());
        System.out.println("response body = " + response.body());

    }

    public static BodyPublisher oMultipartData(Map<Object, Object> data,
                                               String boundary) throws IOException {
        var byteArrays = new ArrayList<byte[]>();
        byte[] separator = ("--" + boundary + "\r\nContent-Disposition: form-data; name=").getBytes(StandardCharsets.UTF_8);
        for (Map.Entry<Object, Object> entry : data.entrySet()) {
            byteArrays.add(separator);
            String x = ("\"" + entry.getKey() + "\"\r\n\r\n" + entry.getValue() + "\r\n");
            byteArrays.add(x.getBytes(StandardCharsets.UTF_8));
        }
        byteArrays
                .add(("--" + boundary + "--").getBytes(StandardCharsets.UTF_8));
        return BodyPublishers.ofByteArrays(byteArrays);
    }

}
```

in the response of java version I see `foo` field:
```
response  = (POST https://httpbin.org/post) 200
response statusCode = 200
response body = {
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {
    "foo": "{\"aaaaa\" : 123 }"
  }, 
  "headers": {
    "Content-Length": "130", 
    "Content-Type": "multipart/form-data; boundary=-------------oiawn4tp89n4e9p5", 
    "Host": "httpbin.org", 
    "User-Agent": "Java-http-client/11.0.7", 
    "X-Amzn-Trace-Id": "Root=1-620cd21e-6a26234d2f674d524972b097"
  }, 
  "json": null, 
  "origin": "......", 
  "url": "https://httpbin.org/post"
}

```


------------------------------------ CLIENT SCALA ----------------------------------
```scala

import java.net.URI
import java.net.http.HttpRequest.BodyPublishers
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.nio.charset.StandardCharsets
import java.util

// build.sbt
// ThisBuild / scalaVersion := "2.13.8"
// libraryDependencies += "org.http4s" %% "http4s-blaze-client" % "0.21.31"

object Test extends App {

  val url = "https://httpbin.org/post"
  
  val data: Map[String, String] = Map("foo" -> """{"aaaaa" : 123}""")
  val boundary = "-------------oiawn4tp89n4e9p5"

  val x = oMultipartData(data, boundary)
  val request = HttpRequest
    .newBuilder()
    .uri(new URI(url))
    .headers("Content-Type", "multipart/form-data boundary=" + boundary)
    .POST(x)
    .build()

  val client = HttpClient.newHttpClient()

  val response = client.send(request, HttpResponse.BodyHandlers.ofString())
  System.out.println("response  = " + response)
  System.out.println("response statusCode = " + response.statusCode())
  System.out.println("response body = " + response.body())

  def oMultipartData(data: Map[String, String], boundary: String): HttpRequest.BodyPublisher = {
    val byteArrays = new util.ArrayList[Array[Byte]]
    val separator  = ("--" + boundary + "\r\nContent-Disposition: form-data; name=").getBytes(StandardCharsets.UTF_8)

    for (entry <- data) {
      val (key, value) = entry
      byteArrays.add(separator)
      val x = s""""$key"\r\n\r\n$value\r\n"""
      byteArrays.add(x.getBytes(StandardCharsets.UTF_8))
    }
    byteArrays.add(("--" + boundary + "--").getBytes(StandardCharsets.UTF_8))
    BodyPublishers.ofByteArrays(byteArrays)
  }

}
```

in the response of scala version the `foo` field inside `form` is empty:
```
response  = (POST https://httpbin.org/post) 200
response statusCode = 200
response body = {
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Content-Length": "130", 
    "Content-Type": "multipart/form-data boundary=-------------oiawn4tp89n4e9p5", 
    "Host": "httpbin.org", 
    "User-Agent": "Java-http-client/11.0.7", 
    "X-Amzn-Trace-Id": "Root=1-620cd08d-5ed81118645a9b6d0eb4d335"
  }, 
  "json": null, 
  "origin": ".....", 
  "url": "https://httpbin.org/post"
}
```

**Comment by gekomad** [comment]
I wrote two wrapper of `java.net.http.HttpClient` in java e in scala.
I sent a multipart POST with field `foo` to https://httpbin.org/post; in response of java version I see the filed `foo` but in scala version I don't see that field.

```
"form": {
    "foo": "{\"aaaaa\" : 123 }"
  }, 
```

--------------------- CLIENT JAVA ------------------------------------

```java
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublisher;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class HttpClientMultipartExample {

    public static void main(String[] args) throws IOException,
            InterruptedException, URISyntaxException {

        String url = "https://httpbin.org/post";

        String boundary = "-------------oiawn4tp89n4e9p5";
        Map<Object, Object> data = new HashMap<>();

        data.put("foo", "{\"aaaaa\" : 123 }");

        BodyPublisher x = oMultipartData(data, boundary);
        HttpRequest request = HttpRequest.newBuilder().uri(new URI(url))
                .headers("Content-Type", "multipart/form-data; boundary=" + boundary).POST(x).build();

        HttpClient client = HttpClient.newHttpClient();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        System.out.println("response  = " + response);
        System.out.println("response statusCode = " + response.statusCode());
        System.out.println("response body = " + response.body());

    }

    public static BodyPublisher oMultipartData(Map<Object, Object> data,
                                               String boundary) throws IOException {
        var byteArrays = new ArrayList<byte[]>();
        byte[] separator = ("--" + boundary + "\r\nContent-Disposition: form-data; name=").getBytes(StandardCharsets.UTF_8);
        for (Map.Entry<Object, Object> entry : data.entrySet()) {
            byteArrays.add(separator);
            String x = ("\"" + entry.getKey() + "\"\r\n\r\n" + entry.getValue() + "\r\n");
            byteArrays.add(x.getBytes(StandardCharsets.UTF_8));
        }
        byteArrays
                .add(("--" + boundary + "--").getBytes(StandardCharsets.UTF_8));
        return BodyPublishers.ofByteArrays(byteArrays);
    }

}
```

in the response of java version I see `foo` field:
```
response  = (POST https://httpbin.org/post) 200
response statusCode = 200
response body = {
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {
    "foo": "{\"aaaaa\" : 123 }"
  }, 
  "headers": {
    "Content-Length": "130", 
    "Content-Type": "multipart/form-data; boundary=-------------oiawn4tp89n4e9p5", 
    "Host": "httpbin.org", 
    "User-Agent": "Java-http-client/11.0.7", 
    "X-Amzn-Trace-Id": "Root=1-620cd21e-6a26234d2f674d524972b097"
  }, 
  "json": null, 
  "origin": "......", 
  "url": "https://httpbin.org/post"
}

```


------------------------------------ CLIENT SCALA ----------------------------------
```scala

import java.net.URI
import java.net.http.HttpRequest.BodyPublishers
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.nio.charset.StandardCharsets
import java.util

// build.sbt
// ThisBuild / scalaVersion := "2.13.8"
// libraryDependencies += "org.http4s" %% "http4s-blaze-client" % "0.21.31"

object Test extends App {

  val url = "https://httpbin.org/post"
  
  val data: Map[String, String] = Map("foo" -> """{"aaaaa" : 123}""")
  val boundary = "-------------oiawn4tp89n4e9p5"

  val x = oMultipartData(data, boundary)
  val request = HttpRequest
    .newBuilder()
    .uri(new URI(url))
    .headers("Content-Type", "multipart/form-data boundary=" + boundary)
    .POST(x)
    .build()

  val client = HttpClient.newHttpClient()

  val response = client.send(request, HttpResponse.BodyHandlers.ofString())
  System.out.println("response  = " + response)
  System.out.println("response statusCode = " + response.statusCode())
  System.out.println("response body = " + response.body())

  def oMultipartData(data: Map[String, String], boundary: String): HttpRequest.BodyPublisher = {
    val byteArrays = new util.ArrayList[Array[Byte]]
    val separator  = ("--" + boundary + "\r\nContent-Disposition: form-data; name=").getBytes(StandardCharsets.UTF_8)

    for (entry <- data) {
      val (key, value) = entry
      byteArrays.add(separator)
      val x = s""""$key"\r\n\r\n$value\r\n"""
      byteArrays.add(x.getBytes(StandardCharsets.UTF_8))
    }
    byteArrays.add(("--" + boundary + "--").getBytes(StandardCharsets.UTF_8))
    BodyPublishers.ofByteArrays(byteArrays)
  }

}
```

in the response of scala version the `foo` field inside `form` is empty:
```
response  = (POST https://httpbin.org/post) 200
response statusCode = 200
response body = {
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Content-Length": "130", 
    "Content-Type": "multipart/form-data boundary=-------------oiawn4tp89n4e9p5", 
    "Host": "httpbin.org", 
    "User-Agent": "Java-http-client/11.0.7", 
    "X-Amzn-Trace-Id": "Root=1-620cd08d-5ed81118645a9b6d0eb4d335"
  }, 
  "json": null, 
  "origin": ".....", 
  "url": "https://httpbin.org/post"
}
```

**Comment by gekomad** [comment]
I wrote two wrapper of `java.net.http.HttpClient` in java e in scala.
I sent a multipart POST with field `foo` to https://httpbin.org/post; in response of java version I see the filed `foo` but in scala version I don't see that field.

```
"form": {
    "foo": "{\"aaaaa\" : 123 }"
  }, 
```

--------------------- CLIENT JAVA ------------------------------------

```java
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublisher;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class HttpClientMultipartExample {

    public static void main(String[] args) throws IOException,
            InterruptedException, URISyntaxException {

        String url = "https://httpbin.org/post";

        String boundary = "-------------oiawn4tp89n4e9p5";
        Map<Object, Object> data = new HashMap<>();

        data.put("foo", "{\"aaaaa\" : 123 }");

        BodyPublisher x = oMultipartData(data, boundary);
        HttpRequest request = HttpRequest.newBuilder().uri(new URI(url))
                .headers("Content-Type", "multipart/form-data; boundary=" + boundary).POST(x).build();

        HttpClient client = HttpClient.newHttpClient();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        System.out.println("response  = " + response);
        System.out.println("response statusCode = " + response.statusCode());
        System.out.println("response body = " + response.body());

    }

    public static BodyPublisher oMultipartData(Map<Object, Object> data,
                                               String boundary) throws IOException {
        var byteArrays = new ArrayList<byte[]>();
        byte[] separator = ("--" + boundary + "\r\nContent-Disposition: form-data; name=").getBytes(StandardCharsets.UTF_8);
        for (Map.Entry<Object, Object> entry : data.entrySet()) {
            byteArrays.add(separator);
            String x = ("\"" + entry.getKey() + "\"\r\n\r\n" + entry.getValue() + "\r\n");
            byteArrays.add(x.getBytes(StandardCharsets.UTF_8));
        }
        byteArrays
                .add(("--" + boundary + "--").getBytes(StandardCharsets.UTF_8));
        return BodyPublishers.ofByteArrays(byteArrays);
    }

}
```

in the response of java version I see `foo` field:
```
response  = (POST https://httpbin.org/post) 200
response statusCode = 200
response body = {
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {
    "foo": "{\"aaaaa\" : 123 }"
  }, 
  "headers": {
    "Content-Length": "130", 
    "Content-Type": "multipart/form-data; boundary=-------------oiawn4tp89n4e9p5", 
    "Host": "httpbin.org", 
    "User-Agent": "Java-http-client/11.0.7", 
    "X-Amzn-Trace-Id": "Root=1-620cd21e-6a26234d2f674d524972b097"
  }, 
  "json": null, 
  "origin": "......", 
  "url": "https://httpbin.org/post"
}

```


------------------------------------ CLIENT SCALA ----------------------------------
```scala

import java.net.URI
import java.net.http.HttpRequest.BodyPublishers
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.nio.charset.StandardCharsets
import java.util

// build.sbt
// ThisBuild / scalaVersion := "2.13.8"

object Test extends App {

  val url = "https://httpbin.org/post"
  
  val data: Map[String, String] = Map("foo" -> """{"aaaaa" : 123}""")
  val boundary = "-------------oiawn4tp89n4e9p5"

  val x = oMultipartData(data, boundary)
  val request = HttpRequest
    .newBuilder()
    .uri(new URI(url))
    .headers("Content-Type", "multipart/form-data boundary=" + boundary)
    .POST(x)
    .build()

  val client = HttpClient.newHttpClient()

  val response = client.send(request, HttpResponse.BodyHandlers.ofString())
  System.out.println("response  = " + response)
  System.out.println("response statusCode = " + response.statusCode())
  System.out.println("response body = " + response.body())

  def oMultipartData(data: Map[String, String], boundary: String): HttpRequest.BodyPublisher = {
    val byteArrays = new util.ArrayList[Array[Byte]]
    val separator  = ("--" + boundary + "\r\nContent-Disposition: form-data; name=").getBytes(StandardCharsets.UTF_8)

    for (entry <- data) {
      val (key, value) = entry
      byteArrays.add(separator)
      val x = s""""$key"\r\n\r\n$value\r\n"""
      byteArrays.add(x.getBytes(StandardCharsets.UTF_8))
    }
    byteArrays.add(("--" + boundary + "--").getBytes(StandardCharsets.UTF_8))
    BodyPublishers.ofByteArrays(byteArrays)
  }

}
```

in the response of scala version the `foo` field inside `form` is empty:
```
response  = (POST https://httpbin.org/post) 200
response statusCode = 200
response body = {
  "args": {}, 
  "data": "", 
  "files": {}, 
  "form": {}, 
  "headers": {
    "Content-Length": "130", 
    "Content-Type": "multipart/form-data boundary=-------------oiawn4tp89n4e9p5", 
    "Host": "httpbin.org", 
    "User-Agent": "Java-http-client/11.0.7", 
    "X-Amzn-Trace-Id": "Root=1-620cd08d-5ed81118645a9b6d0eb4d335"
  }, 
  "json": null, 
  "origin": ".....", 
  "url": "https://httpbin.org/post"
}
```

**Comment by amesgen** [comment]
Did you also force HTTP 2 like you did in your initial example?

**Comment by gekomad** [comment]
I found out the IIS server is behind a Amazon awl Elastic Load Balancing (awselb)

so

standard http4s client works fine:
```
Response(status=200, headers=Headers(Date: Sat, 19 Feb 2022 16:33:18 GMT, Content-Length: 0, Connection: keep-alive, Cache-Control: no-cache, Pragma: no-cache, Expires: -1, Server: Microsoft-IIS/10.0, Access-Control-Allow-Origin: *, X-AspNet-Version: 4.0.30319, X-Powered-By: ASP.NET))
```


With http4s-jdk-http-client I get an error from `awselb`:
```html
<html>
<head><title>400 Bad Request</title></head>
<body>
<center><h1>400 Bad Request</h1></center>
</body>
</html>
```
```
Response(status=400, headers=Headers(connection: close, content-length: 122, content-type: text/html, date: Sat, 19 Feb 2022 16:32:47 GMT, server: awselb/2.0)))
```

**Comment by gekomad** [comment]
**with java version "11.0.7" doesn't work
but with java version "12.0.2" works!**

**Comment by amesgen** [comment]
Ah, then it is an instance of https://github.com/http4s/http4s-jdk-http-client/issues/200, see also this note in the [docs](https://jdk-http-client.http4s.org/):

> TLS 1.3 on Java 11. On Java 11, TLS 1.3 is disabled by default (when using JdkHttpClient.simple). This is a workaround for a spurious bug, see [#200](https://github.com/http4s/http4s-jdk-http-client/issues/200).

In particular, when you create the client via `JdkHttpClient.simple`, it should also work on Java 11.

**Comment by gekomad** [comment]
I'll use jdk 17. Thanks @amesgen

**Comment by epabst** [comment]
If you need to use Java 11, I fixed a similar issue by specifying HTTP 1.1 rather than HTTP 2.
```
HttpClient.newBuilder().version(HttpClient.Version.HTTP_1_1)...
```
